compiler/qsc_data_structures/src/namespaces.rs

```diff
@@ -196,7 +196,75 @@ impl NamespaceTreeRoot {
             .borrow_mut()
             .insert_or_find_namespace(ns.into_iter().peekable(), &mut self.assigner);
 
-        id.expect("empty name should not be passed into namespace insertion")
+        id.expect("empty name checked for above")
+    }
+
+    pub fn insert_or_find_namespace_from_root_with_id(
+        &mut self,
+        mut ns: Vec<Rc<str>>,
+        root: NamespaceId,
+        base_id: NamespaceId,
+    ) {
+        if ns.is_empty() {
+            return;
+        }
+        let (_root_name, root_contents) = self.find_namespace_by_id(&root);
+        // split `ns` into [0..len - 1] and [len - 1]
+        let suffix = ns.split_off(ns.len() - 1)[0].clone();
+        let prefix = ns;
+
+        // if the prefix is empty, we are inserting into the root
+        if prefix.is_empty() {
+            self.insert_with_id(Some(root), base_id, &suffix);
+        } else {
+            let prefix_id = root_contents
+                .borrow_mut()
+                .insert_or_find_namespace(prefix.into_iter().peekable(), &mut self.assigner)
+                .expect("empty name checked for above");
+
+            self.insert_with_id(Some(prefix_id), base_id, &suffix);
+        }
+    }
+
+    /// Each item in this iterator is one namespace. The reason there are multiple paths for it,
```

Comment:
nitpick: for clarity I might phrase this as:

```suggestion
    /// Each item in this iterator is the same, single namespace. The reason there are multiple paths for it,
```